core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/RemoveTrivialFilters.java

```diff
@@ -46,10 +47,23 @@ public Result apply(FilterNode filterNode, Captures captures, Context context)
             return Result.ofPlanNode(filterNode.getSource());
         }
 
-        if (predicate.equals(FALSE_LITERAL) || predicate instanceof NullLiteral) {
+        if (predicate.equals(FALSE_LITERAL) || isEffectivelyNull(predicate)) {
             return Result.ofPlanNode(new ValuesNode(context.getIdAllocator().getNextId(), filterNode.getOutputSymbols(), emptyList()));
         }
 
         return Result.empty();
     }
+
+    private boolean isEffectivelyNull(Expression expression)
+    {
+        if (expression instanceof NullLiteral) {
+            return true;
```

Comment:
This is unreachable, right? Every filter node predicate should be an expression of boolean type, and null literal is not of that type.

Reply:
I'm not sure, this check already existed before.
The following query is valid:
```
trino:sf1000> SELECT * FROM (VALUES '1') WHERE null;
 _col0 
-------
(0 rows)
```
I don't know if we ever create a filter node for a null predicate that is removed by this rule or not, but I don't see a reason why it can't be possible

Reply:
> The following query is valid:
> 
> ```
> trino:sf1000> SELECT * FROM (VALUES '1') WHERE null;
>  _col0 
> -------
> (0 rows)
> ```

it should have null literal coerced to boolean, so it should be `Cast(NullLiteral)`

> I don't know if we ever create a filter node for a null predicate that is removed by this rule or not

i'd assume we do!

but a predicate should be of type boolean!

Reply:
https://github.com/trinodb/trino/pull/15744

Reply:
> it should have null literal coerced to boolean, so it should be `Cast(NullLiteral)`

I am still convinced i was right, but it appears the code was not doing that.

Regular FilterNode and row filters lacked coercion application.
This was found by https://github.com/trinodb/trino/pull/15744 and will be fixed.